### DISCUSSION TITLE
Transparent hugepages configuration

### DISCUSSION THREAD

**Discussion by psact** [bottlerocket-os/bottlerocket#1989] [Q&A] [answered]
Is there a recommended way to modify THPs in bottlerocket? Typically it's done via sysfs, but I don't see anything in the docs that mentions manipulating that.
https://www.kernel.org/doc/html/latest/admin-guide/mm/transhuge.html#thp-sysfs

There's the brute-force way of doing it with pods/daemonsets, but maybe I'm missing a way to pass a kernel arg in the API somewhere.
https://stackoverflow.com/questions/45279477/disable-transparent-huge-pages-from-kubernetes

**Comment by cbgbt** [comment, marked as answer]
Currently, I think the best way to do this would be with a [bootstrap container](https://github.com/bottlerocket-os/bottlerocket#bootstrap-containers-settings). You could do something like this:

```
# In bootstrap container
echo always > /.bottlerocket/rootfs/sys/kernel/mm/transparent_hugepage/enabled
```

I do think it would be ideal to model this with the settings API though. I've opened #1993 as a feature request to support that.

**Reply by psact** [reply]
Do you guys host a docker container that exec's user-data?  I created one and published it to dockerhub, but most people don't want to pull one from some rando like me.

Dockerfile:
```
FROM alpine
ENTRYPOINT ["sh", "/.bottlerocket/bootstrap-containers/current/user-data"]
```
Bootstrap in userdata:
```
[settings.bootstrap-containers.thp]
source = "docker.io/psact/bottlerocket-bootstrap:alpine-latest"
mode = "always"
essential = true
# echo 'echo never > /.bottlerocket/rootfs/sys/kernel/mm/transparent_hugepage/enabled
# echo never > /.bottlerocket/rootfs/sys/kernel/mm/transparent_hugepage/defrag' | base64 -w0
user-data = "ZWNobyBuZXZlciA+IC8uYm90dGxlcm9ja2V0L3Jvb3Rmcy9zeXMva2VybmVsL21tL3RyYW5zcGFyZW50X2h1Z2VwYWdlL2VuYWJsZWQKZWNobyBuZXZlciA+IC8uYm90dGxlcm9ja2V0L3Jvb3Rmcy9zeXMva2VybmVsL21tL3RyYW5zcGFyZW50X2h1Z2VwYWdlL2RlZnJhZwo="
```

**Reply by psact** [reply]
And thanks for all the good work on bottlerocket, much easier than customizing shell scripts in userdata :+1:

**Reply by cbgbt** [reply]
I really like the idea for running a script specified in `user-data`. I've opened #2010 to keep track of that idea. Thanks!

**Reply by James-Quigley** [reply]
I built upon @psact work and made the image multi-arch with support for arm64 as well as amd64. Its published on dockerhub as `oscarhealth/bottlerocket-bootstrap-public:1.0.0` [link](https://hub.docker.com/r/oscarhealth/bottlerocket-bootstrap-public/tags)

**Comment by markusboehme** [comment]
For transparent huge page support, an alternative to using a bootstrap container is to add a kernel command line parameter via user data. While this does get away without an extra container image, it does require a reboot to take effect (which can be automated away with `reboot-to-reconcile`), so there's a trade-off to be made here:

```toml
[settings.boot]
reboot-to-reconcile = true

[settings.boot.kernel]
transparent_hugepage = ["always"]  # or "never" or "madvise"
```

**Reply by andyndang** [reply]
~~This should be the marked answer! So much easier than running a bootstrap action!!~~

NVM. Checked this but it didn't work since `/sys/kernel/mm/transparent_hugepage/enabled` is set to `never` but thp values are still not 0

**Reply by markusboehme** [reply]
Which THP values are you referring to? Please provide some detail on what it is you tested. Note that any THPs already allocated will not be split again just by disabling THPs.
